### PR TITLE
BUG: Fix copy semantics in ``__array__``

### LINKED ISSUE #57739: COMPAT: Utilize `copy` keyword in `__array__`
https://github.com/pandas-dev/pandas/pull/57172 add a copy keyword that will be passed in with `__array__` in Numpy 2.0. Currently it is unused but it would be useful to respect that option when passed

### LINKED ISSUE #59614: BUG: `numpy.ma.fix_invalid` makes changes in-place in numpy 2.1.0 even with `copy=True`
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
>>> import pandas as pd, numpy as np
>>> pd.__version__; np.__version__
'2.2.2'
'2.1.0'
>>> my_series = pd.Series([1.0, 2.0, np.nan, 0.0, 1.0])
>>> my_series
0    1.0
1    2.0
2    NaN
3    0.0
4    1.0
dtype: float64
>>> np.ma.fix_invalid(my_series)
masked_array(data=[1.0, 2.0, --, 0.0, 1.0],
             mask=[False, False,  True, False, False],
       fill_value=1e+20)
>>> my_series
0    1.000000e+00
1    2.000000e+00
2    1.000000e+20
3    0.000000e+00
4    1.000000e+00
dtype: float64
```

### Issue Description

Copying the description of: https://github.com/numpy/numpy/issues/27253

`numpy.ma.fix_invalid` behaves differently between NumPy 2.1.0 and NumPy 2.0.0. Specifically, when passing a pandas Series containing a `numpy.nan` value, `numpy.ma.fix_invalid` now makes changes in-place, even if the `copy` argument is set to its default value of `True`. This issue occurs only with pandas Series, not with NumPy arrays, for example.

### Expected Behavior

```
>>> pd.__version__; np.__version__
'2.2.2'
'2.0.0'
>>> 
>>> my_series = pd.Series([1.0, 2.0, np.nan, 0.0, 1.0])
>>> my_series
0    1.0
1    2.0
2    NaN
3    0.0
4    1.0
dtype: float64
>>> np.ma.fix_invalid(my_series)
masked_array(data=[1.0, 2.0, --, 0.0, 1.0],
             mask=[False, False,  True, False, False],
       fill_value=1e+20)
>>> my_series
0    1.0
1    2.0
2    NaN
3    0.0
4    1.0
dtype: float64
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : d9cdd2ee5a58015ef6f4d15c7226110c9aab8140
python                : 3.11.9.final.0
python-bits           : 64
OS                    : Darwin
OS-release            : 23.5.0
Version               : Darwin Kernel Version 23.5.0: Wed May  1 20:13:18 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6030
machine               : arm64
processor             : arm
byteorder             : little
LC_ALL                : None
LANG                  : None
LOCALE                : None.UTF-8

pandas                : 2.2.2
numpy                 : 2.1.0
pytz                  : 2024.1
dateutil              : 2.9.0
setuptools            : 69.5.1
pip                   : 24.0
Cython                : None
pytest                : 8.3.1
hypothesis            : 6.108.4
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : 3.1.4
IPython               : 8.26.0
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : None
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : 2024.6.1
gcsfs                 : None
matplotlib            : None
numba                 : None
numexpr               : None
odfpy                 : None
openpyxl              : None
pandas_gbq            : None
pyarrow               : 17.0.0
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : None
sqlalchemy            : None
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2024.1
qtpy                  : None
pyqt5                 : None

</details>

### LINKED ISSUE #59932: BUG: np.nan_to_num change pandas DataFrame in place when copy = True
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import numpy as np
import pandas as pd

x = pd.DataFrame([np.nan])
np.nan_to_num(x, copy=True)
print(x)
```

### Issue Description

Result x is changed to 0.0 even when nan_to_num is called with copy=True.

This issue is originally reported to numpy [here](https://github.com/numpy/numpy/issues/27487). Looks like pandas ignores `copy=True` NumPy passes?

### Expected Behavior

x should not be changed when copy=True.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : 0691c5cf90477d3503834d983f69350f250a6ff7
python                : 3.11.8
python-bits           : 64
OS                    : Linux
OS-release            : 5.10.0-smp-1100.466.602.1
Version               : #1 [v5.10.0-1100.466.602.1] SMP @1708156698
machine               : x86_64
processor             : 
byteorder             : little
LC_ALL                : None
LANG                  : None
LOCALE                : en_US.UTF-8

pandas                : 2.2.3
numpy                 : 2.1.1
pytz                  : 2024.1
dateutil              : 2.8.1
pip                   : None
Cython                : None
sphinx                : None
IPython               : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : None
blosc                 : None
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
html5lib              : None
hypothesis            : None
gcsfs                 : None
jinja2                : None
lxml.etree            : 5.3.0
matplotlib            : None
numba                 : None
numexpr               : 2.8.7
odfpy                 : None
openpyxl              : 3.1.2
pandas_gbq            : None
psycopg2              : None
pymysql               : None
pyarrow               : None
pyreadstat            : None
pytest                : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : 1.13.1
sqlalchemy            : None
tables                : None
tabulate              : None
xarray                : None
xlrd                  : 2.0.1
xlsxwriter            : None
zstandard             : None
tzdata                : None
qtpy                  : None
pyqt5                 : None

</details>

### PR DESCRIPTION
**Pull request by seberg**
This should fix the semantics of ``__array__``.  While rejecting ``copy=False`` is OK even if unnecessary, ``copy=True`` should never have been ignored and is dangerous.

Closes gh-57739, closes gh-59932, closes https://github.com/pandas-dev/pandas/issues/59614

- [x] Still needs new tests.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

----

I have to figure out the tests.  I think a test in form of:
```
arr1 = np.asarray(obj, copy=True)
arr2 = np.asarray(obj, copy=True)
assert not np.may_share_memory(arr1, arr2)
# Check that without copy always works:
assert_array_equal(arr1, np.asarray(obj))

if np_ver < 2:
    return  # copy=False semantics not supported

try:
   arr1 = np.asarray(obj, copy=False)
except ValueError:
   return  # An error is acceptable for `copy=False`

# If no error is given, multiple returns must be views:
arr2 = np.asarray(obj, copy=False)
assert np.may_share_memory(arr1, arr2)
```
will work nicely.  But, right now I am not sure if there is a convenient pattern/parametrization to steal to cover all of the `__array__` implementations here.

### CODE REVIEW

File: pandas/core/arrays/categorical.py [hunk 1]
```diff
@@ -1686,13 +1687,20 @@ def __array__(
...
         array(['a', 'b'], dtype=object)
         """
+        if copy is False:
+            raise ValueError(
+                "Unable to avoid copy while creating an array as requested."
+            )
+
         ret = take_nd(self.categories._values, self._codes)
-        if dtype and np.dtype(dtype) != self.categories.dtype:
-            return np.asarray(ret, dtype)
```
Comment (seberg) [inline pandas/core/arrays/categorical.py:1691]: I did not understand why this is needed.  If dtypes match, NumPy should make it a no-op?  If dtype is None, it is the same as not passing.

**Comment by seberg** [comment]
FWIW, if anyone knows how to add decent tests for this, pushing to the PR is appreciated.  Otherwise, I may try to figure out where to add at least tests for a few of these paths.

**Comment by chaoyihu** [comment]
Hi, Glad to see a PR addressing #59932 ! I was trying to reproduce that issue but couldn't install NumPy >= 2.0 with the main branch, which seems to be because the `environment.yml` on the current main branch requires  `numpy<2`. How did you resolve it?

And by the way, I asked about the bug fix release on Slack:

> Could this be released as a bug fix for v2.x?

Quoting the reply from @rhshadrach:

> I think there won't be another 2.2.x release; the next release will be 2.3.0.

**Comment by seberg** [comment]
Found a test that seemed reasonably to expand.  I doubt it covers _all_ the branches changed, though (and no, I don't love expanding tests, but I was lazy to split out the parametrization).

**Comment by seberg** [comment]
Ping @mroeschke just because it's been a while that it was idle without any tests.  Not sure if this needs a milestoned or a  release note.

File: pandas/core/arrays/categorical.py [hunk 2]
```diff
@@ -1686,13 +1687,20 @@ def __array__(
         >>> np.asarray(cat)
         array(['a', 'b'], dtype=object)
         """
+        if copy is False:
```
Comment (mroeschke) [inline pandas/core/arrays/categorical.py:1690]: If `lib.is_range_indexer(self._codes)` i.e. the `self.categories._values` are all unique then, a copy could be avoided?

**Comment by mroeschke** [comment]
Sorry for the delay. Overall looks good! Could you add a note to the `v2.3.0.rst` release notes?

File: pandas/core/arrays/arrow/array.py [hunk 3]
```diff
@@ -667,7 +667,15 @@ def __array__(
...
     ) -> np.ndarray:
         """Correctly construct numpy arrays when passed to `np.asarray()`."""
-        return self.to_numpy(dtype=dtype)
+        if copy is False:
+            # TODO: By using `zero_copy_only` it may be possible to implement this
+            raise ValueError(
+                "Unable to avoid copy while creating an array as requested."
+            )
+        elif copy is None:
+            copy = False  # The NumPy copy=False meaning is different here.
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/arrow/array.py:676]: What does this comment exactly mean?

File: pandas/core/arrays/categorical.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/categorical.py:1691]: Yeah, also not sure why this is here

File: pandas/core/arrays/datetimelike.py [hunk 4]
```diff
@@ -359,7 +359,14 @@ def __array__(
...
         if is_object_dtype(dtype):
+            if copy is False:
+                raise ValueError(
+                    "Unable to avoid copy while creating an array as requested."
+                )
             return np.array(list(self), dtype=object)
+
+        if copy is True:
+            return np.array(self._ndarray, dtype=dtype)
         return self._ndarray
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/datetimelike.py:370]: Does it matter much for the numpy side whether we return the ndarray here directly (which might not match the dtype), or already do `return np.asarray(self._ndarray, dtype=dtype)` on our side?

File: pandas/core/arrays/period.py [hunk 5]
```diff
@@ -389,6 +389,11 @@ def freqstr(self) -> str:
     def __array__(
         self, dtype: NpDtype | None = None, copy: bool | None = None
     ) -> np.ndarray:
+        if copy is False:
+            raise ValueError(
+                "Unable to avoid copy while creating an array as requested."
+            )
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/period.py:395]: Strictly speaking, the 

```
        if dtype == "i8":
            return self.asi8
```

just below is zero copy, so a `copy=False` is possible in that case

Comment (jorisvandenbossche) [reply pandas/core/arrays/period.py:395]: Ah, and more importantly, that means that that branch is not honoring `copy=True`

File: pandas/core/arrays/sparse/array.py [hunk 6]
```diff
@@ -547,6 +547,11 @@ def from_spmatrix(cls, data: spmatrix) -> Self:
...
         self, dtype: NpDtype | None = None, copy: bool | None = None
     ) -> np.ndarray:
+        if copy is False:
+            raise ValueError(
+                "Unable to avoid copy while creating an array as requested."
+            )
+
         fill_value = self.fill_value
 
         if self.sp_index.ngaps == 0:
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/sparse/array.py:557]: corner case, but also this branch is returning underlying values and so is not honoring a `copy=True`

File: pandas/core/generic.py [hunk 7]
```diff
@@ -2015,8 +2015,17 @@ def __array__(
...
-        arr = np.asarray(values, dtype=dtype)
-        if astype_is_view(values.dtype, arr.dtype) and self._mgr.is_single_block:
+        if copy is None:
+            # Note: branch avoids `copy=None` for NumPy 1.x support
+            arr = np.asarray(values, dtype=dtype)
+        else:
+            arr = np.array(values, dtype=dtype, copy=copy)
+
+        if (
+            copy is not False
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py:2025]: and also not `True`, because then it will already have copied in `arr = np.array(values, dtype=dtype, copy=copy)` above?

File: pandas/core/indexes/multi.py [hunk 8]
```diff
@@ -1391,6 +1391,9 @@ def copy(  # type: ignore[override]
 
     def __array__(self, dtype=None, copy=None) -> np.ndarray:
         """the array interface, return my values"""
+        if copy is True:
+            # Note: branch avoids `copy=None` for NumPy 1.x support
+            return np.array(self.values, dtype=dtype, copy=copy)
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/multi.py:1396]:
```suggestion
```

This is unnecessary here, because `MultiIndex.values` is already a customly build array (see `def _values(self)` a bit higher up in this file)

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks a lot @seberg for working in this! 

Added some more comments (some corner cases (like period converted to int being zero-copy) might need additional specific tests, but also happy to work on those)

File: pandas/core/arrays/datetimelike.py (hunk 4 above)
Comment (seberg) [reply pandas/core/arrays/datetimelike.py:370]: No, NumPy will deal with dtype mismatch.  It's been historically more of a hint and not a requirement.

File: pandas/core/arrays/period.py (hunk 5 above)
Comment (seberg) [reply pandas/core/arrays/period.py:395]: Will look into this and other fixups tomorrow.  Right, wrong `copy=True` is a problem, I definitely err'd on the error side for `copy=False`, so let me change it here (as well).

File: pandas/core/generic.py (hunk 7 above)
Comment (seberg) [reply pandas/core/generic.py:2025]: We pass on `copy=copy`, so we can trust the semantics being correct until here.

Comment (seberg) [reply pandas/core/generic.py:2025]: Hmmmm, but I think I got it wrong...  It should rather say, `copy is not True`?  Because if we force a copy it doesn't seem sensible to create a read-only view.

(Have to see if I can write a test for these corrections, it is clearly trickier then just err'ing on the "raise an error" side!)

Comment (jorisvandenbossche) [reply pandas/core/generic.py:2025]: Yes, that is what I meant

File: pandas/core/indexes/multi.py (hunk 8 above)
Comment (seberg) [reply pandas/core/indexes/multi.py:1396]: Ouch, right, in that case we need to raise for ``copy=False`!

File: pandas/core/arrays/categorical.py (hunk 2 above)
Comment (seberg) [reply pandas/core/arrays/categorical.py:1690]: Probably, but I left it out, it doesn't seem vital to try and improve it.  Also fixed things up, because `take_nd` should presumably always return a copy.

**Comment by seberg** [comment]
Pushed fixes based on the review, thanks; it's quite tricky to know whether or not a copy may have been made.

Which also means that expandign the tests would be great (with the new changes also with `dtype=` which I am not sure how well it works in the current test).

@jorisvandenbossche would be great if you can look into tests (I can review), since I have to hunt/think a bit more about how to build examples.  But if I don't see movement in the next days, I'll hopefully dive in.

**Comment by jorisvandenbossche** [comment]
@seberg thanks for the update! I added a bunch of additional tests, that I think should now cover the corner cases for which you updated the code (i.e. PeriodArray and SparseArray corner cases that allow copy=False, MultiIndex never allowing copy=False). 
Also expanded the existing test a bit and added a basic base extension array test.

File: pandas/core/arrays/masked.py [hunk 9]
```diff
@@ -507,7 +507,7 @@ def to_numpy(
         else:
             with warnings.catch_warnings():
                 warnings.filterwarnings("ignore", category=RuntimeWarning)
-                data = self._data.astype(dtype, copy=copy)
+                data = np.array(self._data, dtype=dtype, copy=copy)
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/masked.py:510]: Made this change together with allowing to get here with `copy=False`, to ensure if `copy=False` is not actually possible depending on the passed `dtype`, it still properly errors

File: pandas/tests/indexes/multi/test_conversion.py [hunk 10]
```diff
@@ -16,6 +18,40 @@ def test_to_numpy(idx):
...
+        ("qux", "one"),
+        ("qux", "two"),
+    ]
+    tm.assert_numpy_array_equal(result, expected)
+
+    # it always gives a copy by default, but the values are cached, so results
+    # are still sharing memory
+    result_copy1 = np.asarray(idx)
+    result_copy2 = np.asarray(idx)
+    assert np.may_share_memory(result_copy1, result_copy2)
```
Comment (jorisvandenbossche) [inline pandas/tests/indexes/multi/test_conversion.py:39]: This might be an unfortunate side effect of our caching, because a user might expect here always to get a new object. But at least when explicitly passing `copy=True`, then it appears to do the right thing. But is this numpy that does that? (still copy even if `copy=True` was passed down to `__array__`?) We should probably not keep relying on that, and do an explicit copy on our side in case of `copy=True`?

File: pandas/core/arrays/masked.py (hunk 9 above)
Comment (seberg) [reply pandas/core/arrays/masked.py:510]: This change (together with the `self._hasna` change below) make sense to me.  But unfortunately, `to_numpy(copy=False)` change the meaning of `copy=False` this way.  And I think this is public API?

Maybe better to special case `_hasna` in the branch below?  (could even just ignore `dtype` with a comment that NumPy is fine with that).

File: pandas/tests/base/test_conversion.py [hunk 11]
```diff
@@ -374,15 +380,16 @@ def test_to_numpy(arr, expected, index_or_series_or_array, request):
...
 
-    try:
-        result_nocopy1 = np.array(thing, copy=False)
-    except ValueError:
-        # An error is always acceptable for `copy=False`
-        return
+    if not zero_copy:
+        with pytest.raises(ValueError, match="Unable to avoid copy while creating"):
+            # An error is always acceptable for `copy=False`
+            np.array(thing, copy=False)
```
Comment (seberg) [inline pandas/tests/base/test_conversion.py:386]: Very nice to make the test specific!

File: pandas/tests/extension/base/interface.py [hunk 12]
```diff
@@ -71,6 +73,25 @@ def test_array_interface(self, data):
...
+
+        if not np_version_gt2:
+            # copy=False semantics are only supported in NumPy>=2.
+            return
+
+        try:
+            result_nocopy1 = np.array(data, copy=False)
+        except ValueError:
+            # An error is always acceptable for `copy=False`
+            return
```
Comment (seberg) [inline pandas/tests/extension/base/interface.py:89]: You can make this test specific presumably.

**Review by seberg — COMMENTED** [review]
Thanks a lot, the test additions look good to me modulo a nitpick.

I am worried about the change of the meaning for `.to_numpy(copy=False)`.  But maybe this is private API?

If this is private API, then it is OK, but:
* The docs should be changed and both branches should behave the same (i.e. raise an error in the first branch if `copy=False`).
* The default should be `copy=None`.

If not, we can move that special case to the `__array__`.

File: pandas/tests/extension/base/interface.py (hunk 12 above)
Comment (jorisvandenbossche) [reply pandas/tests/extension/base/interface.py:89]: For this specific case it is harder to make it specific, because we don't know the exact array object and its expected semantics (this is eg also used by external packages implementing an ExtensionArray)

File: pandas/core/arrays/masked.py (hunk 9 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/masked.py:510]: Yeah, will have to update this to not change `to_numpy(copy=False)`, as that is indeed public API we want to keep working that way.

File: pandas/tests/indexes/multi/test_conversion.py (hunk 10 above)
Comment (jorisvandenbossche) [reply pandas/tests/indexes/multi/test_conversion.py:39]: It seems I am testing locally with numpy 2.0.2 (so not 1.x, but also not latest 2.1.x or main)

Comment (seberg) [reply pandas/tests/indexes/multi/test_conversion.py:39]: Ah, maybe there were fixes with 2.1.

**Comment by seberg** [comment]
Thanks Joris.  With the last two commits, I think this should be good to go in.

**Review by mroeschke — APPROVED** [review]

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
The whatsnew note can probably be improved, but going to merge this already so I can start on backporting the code to 2.3.x.

Thanks @seberg!

**Comment by jorisvandenbossche** [comment]
Manual backport -> https://github.com/pandas-dev/pandas/pull/60189

File: pandas/core/series.py [hunk 13]
```diff
@@ -879,8 +879,15 @@ def __array__(
...
-        if astype_is_view(values.dtype, arr.dtype):
+        if copy is None:
+            # Note: branch avoids `copy=None` for NumPy 1.x support
+            arr = np.asarray(values, dtype=dtype)
+        else:
+            arr = np.array(values, dtype=dtype, copy=copy)
+
+        if copy is True:
+            return arr
+        if copy is False or astype_is_view(values.dtype, arr.dtype):
```
Comment (jorisvandenbossche) [inline pandas/core/series.py:890]: One more issue I noticed while doing the backport: for a similar case in generic.py we changed the `copy is False` to `copy is not True`, which I think we should have done here as well.

Will try to come up with a test case that would fail because of this, and do a follow-up tomorrow.

Comment (seberg) [reply pandas/core/series.py:890]: Thanks, ouch.  This is clearly hard to get right without tests :/.

The release snippet looks fine to me.

Comment (jorisvandenbossche) [reply pandas/core/series.py:890]: It also only matters for the `copy=None` case, which AFAIU is quite new, and so we didn't cover that when adding the logic here for making the resulting array read-only

Comment (seberg) [reply pandas/core/series.py:890]: Looking closer, I think this is right?  Note that it says `copy is False or ...`.  If `copy is False` we don't have to check whether it is a view (because otherwise it would have errored)?

Comment (jorisvandenbossche) [reply pandas/core/series.py:890]: Ah, that's a good point. I assume the `arr = np.array(values, dtype=dtype, copy=copy)` line above will already have errored in case of `copy=False` if a zero-copy conversion was not possible. 
So indeed the `or` is fine: if `copy=False` we know `arr` is always a view, and for `copy=None` (the one remaining case) we have `astype_is_view` to check it.

Maybe it's then in `generic.py` that we could change the `copy is not True and ..` to `copy is False or ..` (just to use a similar pattern in both cases)

Comment (jorisvandenbossche) [reply pandas/core/series.py:890]: Looking once more at it, we actually already have tests for this, because we test that we get a read-only array using `np.asarray(ser)`, which essentially means passing `copy=None`. 

I expanded the test with `np.array(ser, copy=False)` to also explicitly cover the `copy=False` case (and ensure this branch setting writeable to False is covered for that case) -> https://github.com/pandas-dev/pandas/pull/60191

**Comment by jorisvandenbossche** [comment]
@seberg I am seeing some failures in the geopandas tests using pandas dev because of this change (https://github.com/geopandas/geopandas/actions/runs/11688928152/job/32550520328?pr=3455)

Essentially geopandas is using `np.array(pandas_object, copy=False)` in a few places. And in the past this has always worked, just silently returning a copy (I assume we just passed `copy=False` in the idea of avoiding a copy if not needed, it's not that the code relied on it returning a view). 
I think for numpy itself you made this behaviour change of `np.array` (to start raising an error) on 2.0? Did you warn about it in advance?

I am wondering if we should leave the `copy=False`-raising-error part of the changes for pandas 3.0, instead of including that in pandas 2.3.

The `copy=True` part of the change (actually copying when asked) is I think the main bug fix we definitely should include in 2.3, as far as I understand (I think the issues people have been reporting in https://github.com/pandas-dev/pandas/issues/57739 would be resolved by this part)

**Comment by seberg** [comment]
No, there was no warning about it, it was a hard change in 2.0...

You'll be the better judge for pandas users here.  It is wrong to not raise, but it does seem worse to backport `copy=False` for sure, since `copy=False` will be exceedingly rare anyway in code.
If this was right from the start geopandas would have had to fix this as part of NumPy 2 transition, but that ship has sailed.

So I guess, you may only want to backport the `copy=True` fix, since it seems important, but not touch `copy=False` even if it is a bug.
If the next release is 3.0, I guess that is good enough to just do the change without deprecation then.

**Comment by simonjayhawkins** [comment]
> I am seeing some failures in the geopandas tests using pandas dev because of this change

@jorisvandenbossche now that this is merged and backported should we have a new issue for this and linked from the 2.3 release discussion for greater visibility?

> I am wondering if we should leave the copy=False-raising-error part of the changes for pandas 3.0, instead of including that in pandas 2.3.

Seems reasonable to me that we probably don't want to introduce breaking changes in 2.3

**Comment by jorisvandenbossche** [comment]
Yes, opened https://github.com/pandas-dev/pandas/issues/60340 and labeled it for 2.3.0
